File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-18.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1
        while right - left > 1:
            print(f"{left=}, {right=}, {nums[left]=}, {nums[right]=}")
            if nums[left] == target:
                return left
            elif nums[right] == target:
                return right

            mid = (left + right) // 2

            if nums[mid] == target:
                return mid

            elif nums[left] < target and target < nums[mid]:
                right = mid

            elif nums[right] > target and target > nums[mid]:
                left = mid

            elif nums[mid] > target and nums[left] < nums[mid]:
                left = mid

            elif nums[mid] < target and nums[right] > nums[mid]:
                right = mid

            elif nums[mid] < target:
                left = mid

            else:
                right = mid

        if nums[left] == target:
                return left
        elif nums[right] == target:
            return right

        return -1 

File: Data Structures & Algorithms/is-palindrome/submission-18.ts
class Solution {
    /**
     * @param {string} s
     * @return {boolean}
     */
  isPalindrome(s: string): boolean {
      s = s.toLowerCase()
      let left: number = 0;
      let right = s.length - 1;

      while (left < right) {
        if (left >= s.length) break;
        else if (right < 0) break;
        else if (!this.isAlphaNum(s[left])) left++;
        else if (!this.isAlphaNum(s[right])) right--;
        else if (s[left] == s[right]) {
          left++; right--;
        }
        else return false;
      }
    return true;
  }

  isAlphaNum(c: string): boolean {
    if (c.valueOf() >= 'a' && c.valueOf() <= 'z') return true;
    if (c.valueOf() >= '0' && c.valueOf() <= '9') return true;
    return false;
  }
}
